File: frontend/src/lib/schemas/printed-transfer.ts
/**
 * Zod schemas for the Printed Transfers (estampados) tier.
 *
 * Replaces the old `print-stock` schemas. Mirrors the Pydantic shapes from
 * `backend/src/schemas/printed_transfer.py`. A `PrintedTransfer` is keyed by
 * `(print_design, side)` via a real FK — NOT the old free-text `product_color`.
 * On-hand is derived live from an append-only ledger
 * (`printed_transfer_movements`): ENTRY and ADJUSTMENT credit, EXIT debits.
 * Quantities are integer counts; `low_stock` is a boolean.
 */

import { z } from "zod";

export const PRINTED_MOVEMENT_KINDS = ["entry", "exit", "adjustment"] as const;
export const printedMovementKindSchema = z.enum(PRINTED_MOVEMENT_KINDS);
export type PrintedMovementKind = z.infer<typeof printedMovementKindSchema>;

export const PRINT_SIDES = ["front", "back"] as const;
export const printSideSchema = z.enum(PRINT_SIDES);
export type PrintSide = z.infer<typeof printSideSchema>;

const pageMeta = {
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
  has_more: z.boolean(),
};

// ---------- Nested DTOs ----------

export const printDesignMiniSchema = z.object({
  id: z.string(),
  code: z.string(),
  name: z.string(),
  image_url: z.string().nullable().optional(),
});
export type PrintDesignMini = z.infer<typeof printDesignMiniSchema>;

// ---------- Levels (printed transfer x on-hand) ----------

export const printedTransferLevelReadSchema = z.object({
  printed_transfer_id: z.string(),
  print_design_id: z.string(),
  design: printDesignMiniSchema,
  side: printSideSchema,
  min_stock: z.number().int().nullable().optional(),
  on_hand: z.number().int(),
  in_production: z.number().int(),
  low_stock: z.boolean(),
  entries_total: z.number().int(),
  exits_total: z.number().int(),
  last_movement_at: z.string().nullable().optional(),
});
export type PrintedTransferLevelRead = z.infer<typeof printedTransferLevelReadSchema>;

export const printedTransferLevelPageSchema = z.object({
  items: z.array(printedTransferLevelReadSchema),
  ...pageMeta,
});
export type PrintedTransferLevelPage = z.infer<typeof printedTransferLevelPageSchema>;

// ---------- Movements ledger ----------

export const printedMovementReadSchema = z.object({
  id: z.string(),
  printed_transfer_id: z.string(),
  design: printDesignMiniSchema.nullable().optional(),
  side: printSideSchema,
  kind: printedMovementKindSchema,
  quantity: z.number().int(),
  notes: z.string().nullable(),
  created_at: z.string(),
});
export type PrintedMovementRead = z.infer<typeof printedMovementReadSchema>;

export const printedMovementPageSchema = z.object({
  items: z.array(printedMovementReadSchema),
  ...pageMeta,
});
export type PrintedMovementPage = z.infer<typeof printedMovementPageSchema>;

// ---------- Create payloads ----------

export type PrintedTransferCreate = {
  print_design_id: string;
  side: PrintSide;
  min_stock?: number | null;
};

export type PrintedMovementCreate = {
  printed_transfer_id: string;
  kind: PrintedMovementKind;
  quantity: number;
  notes?: string | null;
};

// ---------- Filters ----------

export type PrintedTransferLevelFilters = {
  q?: string;
  print_design_id?: string;
  side?: PrintSide;
  low_stock_only?: boolean;
  page?: number;
  page_size?: number;
};

export type PrintedMovementFilters = {
  printed_transfer_id?: string;
  print_design_id?: string;
  side?: PrintSide;
  kind?: PrintedMovementKind;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
};

File: backend/src/schemas/printed_transfer.py
"""Pydantic schemas for the Printed Transfers (estampados) feature.

Replaces the old ``print_stock`` schemas. A ``PrintedTransfer`` is keyed by
``(print_design, side)`` via a real FK (not the old free-text ``product_color``).
On-hand is derived live from a single append-only ledger
(``printed_transfer_movements``): ENTRY and ADJUSTMENT credit stock, EXIT debits
it. The service aggregates ledger rows into ``PrintedTransferLevelRead`` per
printed transfer (every catalog row is surfaced, even with zero movements);
movement mutations are POST-only and surface as ``PrintedMovementRead``.

Quantities are integer counts. ``low_stock`` is computed per row from the
row-level ``min_stock`` (when set) else the company-wide ``printed`` threshold;
``in_production`` is a typed placeholder (0) until the WIP wiring lands later.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from models.enums import PrintedMovementKind, PrintSide
from schemas._common import Page

# ---------- Catalog create ----------


class PrintedTransferCreate(BaseModel):
    print_design_id: uuid.UUID
    side: PrintSide
    min_stock: int | None = Field(default=None, ge=0)


# ---------- Nested DTOs ----------


class PrintDesignMini(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    image_url: str | None = None


# ---------- Printed transfer x on-hand aggregate ----------


class PrintedTransferLevelRead(BaseModel):
    """A single printed transfer's aggregated on-hand position."""

    printed_transfer_id: uuid.UUID
    print_design_id: uuid.UUID
    design: PrintDesignMini
    side: PrintSide
    min_stock: int | None = None
    on_hand: int
    in_production: int
    low_stock: bool
    entries_total: int
    exits_total: int
    last_movement_at: datetime | None = None


# ---------- Ledger movements ----------


class PrintedMovementCreate(BaseModel):
    printed_transfer_id: uuid.UUID
    kind: PrintedMovementKind = Field(default=PrintedMovementKind.ENTRY)
    quantity: int = Field(gt=0)
    notes: str | None = Field(default=None, max_length=500)


class PrintedMovementRead(BaseModel):
    id: uuid.UUID
    printed_transfer_id: uuid.UUID
    design: PrintDesignMini | None = None
    side: PrintSide
    kind: PrintedMovementKind
    quantity: int
    notes: str | None
    created_at: datetime


# ---------- Filters ----------


class PrintedTransferLevelFilters(BaseModel):
    q: str | None = Field(default=None, max_length=120)
    print_design_id: uuid.UUID | None = None
    side: PrintSide | None = None
    low_stock_only: bool = False


class PrintedMovementFilters(BaseModel):
    printed_transfer_id: uuid.UUID | None = None
    print_design_id: uuid.UUID | None = None
    side: PrintSide | None = None
    kind: PrintedMovementKind | None = None
    date_from: date | None = None
    date_to: date | None = None


# ---------- Page aliases ----------


PrintedTransferLevelPage = Page[PrintedTransferLevelRead]
PrintedMovementPage = Page[PrintedMovementRead]
